### DISCUSSION TITLE
Ports & CrossSections

### DISCUSSION THREAD

**Discussion by thomasdorch** [gdsfactory/gdsfactory#2561] [General]
I'd like to propose that we make a change to the way we manage creating `Ports` from `CrossSections`. At the moment, a `CrossSection` contains individual `Sections`, which each represent some sort of 'signal'. When we extrude the `CrossSection` into a `Component`, we create a `Port` for each `Section` that specifies a `port_name`, indicating a point at which we can connect to that particular signal. When creating this `Port`, we add the `CrossSection` to it, but this `CrossSection` contains information about all of the other `Sections`, rather than just the one we created the `Port` for. This means that if we make 2 parallel waveguides (as in a coupler), and try to create a route from just one of them, using the `Port`'s `CrossSection`, we have to manually create a new `CrossSection` based on only the `Section` we want to route from.

One way to improve this could be to create a new `Pin` object that gets created for each of the `Sections` and stored within the `Port`. In other words, rather than creating a `Port` for each `Section` when extruding a `CrossSection`, we would only create one `Port` and each of the `Sections` get a `Pin` that's stored within that `Port`. This would be helpful in the verifying overlaps for waveguides that contain many layers, where we would be able to check whether a `Port` has all of it's `Pins` connected. 

Here's an example of what I envision a `Pin` to look like:
```
class Pin(BaseModel):
    name: str = "o1"
    layer: LayerSpec
    width: float
    pin_type: str = "optical"
```

I'm very interested to see what y'all think of this or whether there's another better way to store individual `Sections` inside of `Ports` rather than the entire `CrossSection`

**Comment by joamatab** [comment]
This could be part of the next major release that we've been planning for a while
https://github.com/gdsfactory/gdsfactory/issues/2557

@tvt173 
@sebastian-goeldi

**Reply by sebastian-goeldi** [reply]
I am not even sure how that would be necessary with kfactory. As long as everything but the main layer of the crossection can be a static distance from the generated mainlayer, we don't really need crossections.

If it is not a static distance, extrude both and apply the KCellEnclosure then.

**Reply by sebastian-goeldi** [reply]
Also, I want to stress that with the introduction of klayout metadata we cannot use complex structures anymore in the info. On write they will be converted to string (maybe even an error, I am not sure about the info part)
